### DISCUSSION TITLE
Permanent Token

### DISCUSSION THREAD

**Discussion by IzMichael** [pocketbase/pocketbase#342] [Q&A] [answered]
Is there any way for me to have admin tokens be permanent? I don't want to have to login each time. Thanks!

**Comment by ganigeorgiev** [comment, marked as answer]
No, at least not exactly "permanent".

You can change the default admin auth token duration from "Settings > Token options > Admin auth token duration" and set it to its current maximum allowed value of 1 year or 31536000 seconds (I'll consider increasing the maximum in the next release).

**Please note that the token will still be invalidated on admin password change.**

**Comment by calumk** [comment]
@ganigeorgiev 
This seems like a useful feature, but its more related to the API, rather than the actual process of logging in. 

It feels quite dated putting a username/email in the auth sdk, and is probably pretty bad for password reuse etc.

What is the suggested use case for say, a public facing website? 
That the site is created as a "user"? or that the site is created as an "admin"?

if you have two sites, accesing the same data, do you make two admin accounts? 
why should a site, have to have a email associated with it? - a token would be much more practical.

**Reply by ganigeorgiev** [reply]
> It feels quite dated putting a username/email in the auth sdk, and is probably pretty bad for password reuse etc.
> What is the suggested use case for say, a public facing website?
> That the site is created as a "user"? or that the site is created as an "admin"?


I don't understand the above statement.
For public data you don't need to provide authentication token (this is managed through "_Collection settings > API rules_").
For the cases when the collection requires authentication - how would you login admin or user without providing auth credentials?


> if you have two sites, accesing the same data, do you make two admin accounts?
> why should a site, have to have a email associated with it? - a token would be much more practical.

Does it really matter whether an API key/secret or Email/Pass combination is used?
At the end - both are just credentials and serve the same purpose.

We can add support for **fixed API keys** which will work in a similar fashion to an Admin authentication, but this may cause some confusions and I don't really see much benefit in this feature. The only use case where I can think of this would be useful is when you want to query/modify the database without the Collection rules restriction behind the scene, for example when in SSR or in a direct server-server communication. You can do the same with the admin email/pass flow, but it requires you to make a "login" call beforehand.

**Reply by IzMichael** [reply]
As a workaround, would something like `@request.query.token = "LONG TOKEN HERE"` in the collection rules, and then pass the token as a URL query param work?

*syntax might be slightly off, but you get the gist*

**Comment by calumk** [comment]
I get your point, and very aware your time is limited, but I think the problem is that the email/pass flow  just _feels_ wrong, for API stuff.

It suggest that it is a person accessing the data, where there are many possibilities where it is not a person, but a service, or worker, or app, or script etc.

Lets say i create 5 or 6 scripts, or small apps, or websites, or.... etc. that access this same data.
Or a native mobile app (for example) and a web-app
Should each one have a dummy admin account created?
Should i just use create one admin account and they can all share?
Should i have to just make up email addresses?  app1@fake.com, app2@fake.com

Of course, In theory, there is not really a practical difference between an email address / password, and a key/token - but my suggestion would be that the admin account could create api keys, that could be paused/revoked, and named after each specific app that is interfacing with pocketbase

**Reply by ganigeorgiev** [reply]
> Lets say i create 5 or 6 scripts, or small apps, or websites, or.... etc. that access this same data. Or a native mobile app (for example) and a web-app

Admins are practically developer/system accounts. You shouldn't hardcode any credentials in any frontend application (including native mobile apps).

> Should each one have a dummy admin account created?
> Should i just use create one admin account and they can all share?
> Should i have to just make up email addresses? app1@fake.com, app2@fake.com

All admins act as super-admin, meaning that API rules don't apply for them and they can read/modify everything.
It's up to you whether you want to use different admin accounts for each external service (the same is true when you are using an API key).

I understand the use case when only server-server communication is concerned (eg. as in your example service or worker) and as I said, we can implement in addition to the admin email/pass flow **fixed API keys** but this may cause some ambiguities with the API usage - we practically will have 3 `Authorization` schemas and will have to explain to the end users, that only the `Users` schema is Collection API rules constrained. And this could get even more messier when PocketBase is used as framework.

I really want to avoid complicating the internals and I don't want to provide multiple ways of doing more-or-less the same thing, and therefore at least for v1.0.0 I'm not planning to wok on API keys support.

**Comment by VladimirCores** [comment]
Not sure if it is right place to ask my question, but looks like relative to "permanent token" - Is there any way I can get public JWT key to add it to another service (https://docs.couchdb.org/en/latest/api/server/authn.html#jwt-authentication)?

**Reply by ganigeorgiev** [reply]
You can get the last authenticated JWT token from the AuthStore by doing something like `client.authStore.token` but this token is not permanent (the duration could be controlled from Admin UI > Settings > Token Options). I'm not sure whether it could be used with the CouchDB Auth but you could give it a try.

**Reply by VladimirCores** [reply]
The thing is that the token should be persistent because it is added to CouchDB settings and used to verify all requests with JWT.

**Reply by ganigeorgiev** [reply]
After v0.8, I'll revisit @calumk proposal for fixed API keys (the authorization schemas are no longer required there because the token type is resolved from the payload).

**Reply by eqoram** [reply]
Is there any update about the permanent API key? :)

**Reply by ganigeorgiev** [reply]
@eqoram Nothing has changed. It exists in my todo to consider it but it is left aside as other tasks are with higher priority.

**Comment by ganigeorgiev** [comment]
_Just for info in case someone stumble here - with [JS SDK v0.18.0-rc2](https://www.npmjs.com/package/pocketbase/v/0.18.0-rc2) release there is now a new experimental `autoRefreshThreshold` option that will take care to maintain a "persistent" admin auth state. For more details and example you can check https://github.com/pocketbase/pocketbase/issues/3118#issuecomment-1701180778._
